### ISSUE TITLE
Solution to avoid caching old resources at upgrade - revisited

### ISSUE THREAD

**Issue by philwareham** [textpattern/textpattern#1038]
### Expected behaviour

The system should load in the latest version of `textpattern.js` and other resources when upgrading a Textpattern installation.

### Actual behaviour

We see a lot of support questions on this - people upgrade but unbeknownst to them the older JavaScript and/or other resources are cached and this can cause breakage unto the user force refreshes their cache. We discussed it previously but just wondering if you had thoughts on a solution given newer technologies are available now...

See https://github.com/textpattern/textpattern/issues/291

**Comment by Bloke** [comment]
Agreed, it is a pain. But after much to-and-fro in the attached issue #291 (which is a good read - stick with it to find out whodunnit) we all came to the conclusion that the best option was to leave it the hell alone.

The facility to tag version numbers is in core. We just don't use it yet. If someone wants to take the ideas presented in that thread and serve up a solution that doesn't impact performance every request, then let's try it again.

Otherwise, we'll just live with it.

**Comment by philwareham** [comment]
I wonder if the [Cache API](https://developer.mozilla.org/en-US/docs/Web/API/Cache) has any potential here? Not sure.

**Comment by bloatware** [comment]
Forced [reload](https://developer.mozilla.org/en-US/docs/Web/API/Location/reload) on update (or diag), maybe.

**Comment by bloatware** [comment]
Is this [this easy](https://github.com/textpattern/textpattern/commit/452ede87ab742f975677d2ed2fabff45f591ba2e)? Should work for regular updates, but I'm not sure for dev... at worst, we can include it only in releases.

**Comment by makss** [comment]
As I understand it, this reloads the page itself, and not the javascript file.
We need to reset the browser cache for the file `textpattern.js`.

**Comment by makss** [comment]
Add to `.htaccess`
```
ExpiresActive On
ExpiresByType image/jpg A604800
ExpiresByType image/jpeg A604800
ExpiresByType image/gif A604800
ExpiresByType image/png A604800
ExpiresByType image/x-icon A604800
ExpiresByType text/css A604800
ExpiresByType application/javascript A604800
```

This will allow the browser to cache and it will be easier to see the problem.

**Comment by bloatware** [comment]
> As I understand it, this reloads the page itself, and not the javascript file.

From my tests on FF57 W7, it behaves like Ctrl+F5, reloading all page resources. But I invite everyone to check: save `_update.php`, open the network console and access some backend page.

> This will allow the browser to cache and it will be easier to see the problem.

Generally the problem appears immediately on update, not one week after.

**Comment by phiw13** [comment]
> From my tests on FF57 W7, it behaves like Ctrl+F5, reloading all page resources. But I invite everyone to check: save _update.php, open the network console and access some backend page.

On my Mac, Safari 11 + Firefox, the document is reloaded (in both cases I accessed the Diagnostics panel), but all the resources are still loaded from cache.

![reload](https://user-images.githubusercontent.com/8944809/33416513-533f8e6c-d5e0-11e7-99e2-181644063f59.png)

**Comment by bloatware** [comment]
Sorry @phiw13, I had to disable it on dev because it also removes `setup` folder now. To test, please edit temporarily `txp_diag.php`:

- line 197: `if ($txp_is_dev) {`
- line 203: `//Txp::get('Textpattern\Admin\Tools')->removeFiles(txpath, 'setup');`

Save `_update.php`, test, revert back and report. Thanks!

**Comment by phiw13** [comment]
@bloatware 
Ok Firefox 57 now visibly reloads the page, in the network panel I see that all resources are downloaded (with status 200).

With Safari, accessing the Diagnostics panel there is a visible reload, all resources (except the PHP file and the SVG files embedded in the stylesheet) show a status of 304, and some 180~200B fetched per file. The PHP and the SVG files are status 200.

**Comment by bloatware** [comment]
Bad, Webkit, [very bad](https://www.chromestatus.com/feature/5724282256621568). I don't see a better option than alerting the user on update, then.

**Comment by philwareham** [comment]
Cache API and Service Worker API seem to be able to manipulate the browser cache: https://developers.google.com/web/ilt/pwa/lab-caching-files-with-service-worker

...but the documentation is almost impenetrable for me, so who knows?!

Anyway, if `ctrl + F5`, ` Apple + R` , and `F5` is the only safe option, then that's OK. We'll just have to make that as visible as possible and continue to answer support questions that arise because of this.

Thanks everyone!

**Comment by phiw13** [comment]
For the docs:

* Firefox Mac: `Command-Shift-R` fetches from server (force reload)
* Safari: `Command-Option-R` bypasses memory caches (usually enough to update), a second `Cmd-Opt-R` fetches from server
* Opera / Chrome Mac: `Command-Option-R` but I don’t know all the details.

**Comment by bloatware** [comment]
[Cache API](https://developer.mozilla.org/en-US/docs/Web/API/Cache) is flagged "experimental" and not universally supported, so not for 4.7 anyway. Well, go for a message.

**Comment by phiw13** [comment]
> Bad, Webkit, very bad. I don't see a better option than alerting the user on update, then.

That is Blink (Chrome), not Webkit (Safari)…

I did a second test with Safari, slightly more  realistic: open/save `_update.php` but *also* I modified the stylesheet. This time, accessing Diagnostics completely fetched the freshly updated stylesheet, the whole 100kb+ with status 200.

**Comment by philwareham** [comment]
> Cache API is flagged "experimental" and not universally supported

True, but Edge, Chrome and Firefox support it and other browsers (i.e. Safari) can be bypassed using code like this:

```
// ServiceWorker is a progressive technology. Ignore unsupported browsers
if ('serviceWorker' in navigator) {
  console.log('CLIENT: service worker registration in progress.');
  navigator.serviceWorker.register('https://cdn.css-tricks.com/service-worker.js').then(function() {
    console.log('CLIENT: service worker registration complete.');
  }, function() {
    console.log('CLIENT: service worker registration failure.');
  });
} else {
  console.log('CLIENT: service worker is not supported.');
}
```

Anyway, let's park this issue until at least Textpattern 4.8, when *maybe* it'll be a viable option. At least there is the possibly of a solution in future.

**Comment by bloatware** [comment]
> I did a second test with Safari, slightly more realistic: open/save _update.php but also I modified the stylesheet. This time, accessing Diagnostics completely fetched the freshly updated stylesheet, the whole 100kb+ with status 200.

Yes, even 304 should be fine, so there still is some hope. But I was unable to make Chrome on W7 behave itself, damn Google. Let's put it on "unsupported browsers" list?

**Comment by bloatware** [comment]
Dunno what to do with it. Even if we refresh the cache of the *updating* browser, other browsers will still keep outdated resources. ATM I have introduced `textpattern.version` in `textpattern.js` that must be identical to `txp_version` constant (otherwise one gets an alert on Diagnostics tab). But it is not very reliable if cached `textpattern.js` is really outdated, and we can not add a version to every resource. TBC...

**Comment by philwareham** [comment]
I’d say back out the code. Leave it until we have a clear solution. Leave the issue open.

**Comment by Bloke** [comment]
I say forget it. This is way too much hassle!

**Comment by makss** [comment]
If more than one person works with Textpattern, then this warning does not help much.

I would not do anything, well, or I would do `textpattern.js?v=4.7.0`. The URL will change so the browser will ask for a new version of the file.

**Comment by philwareham** [comment]
The last time we discussed this, the query string approach was unfavourable due to performance impact (Robert surmised around 2-3 milliseconds slower pagevrendering time). Not sure that has changed in last couple of years. FWIW most other web apps use that method.

**Comment by bloatware** [comment]
No, the issue is too annoying, especially in 4.7 which will not work with 4.6 `textpattern.js`. External `v=4.7.0` versioning is an easy solution, but it fills the browser cache with outdated versions and does not help if someone has uploaded a wrong file (this happens too). Leave it with me if you wish, I won't play with Cache API right now, but have enough time to investigate some SRI check. Just need an Oxford English wording atm ;-)

**Comment by bloatware** [comment]
I have tested SRI for js includes, it works just fine, but is currently supported only by FF, Chrome and Safari 11. It requires, however, enabling CORS `Access-Control-Allow-Origin` header on the server, though we can bypass it by serving `textpattern.js` via PHP (with appropriate headers).

Not sure we'll have time to test it before 4.8, just leaving these notes here.

**Comment by bloatware** [comment]
I have created [integrity](https://github.com/textpattern/textpattern/tree/integrity) branch, where `textpattern.js` is served via a basic `server.php` script and imported with `integrity` attribute. In my tests (FF57 W7) the file is served with roughly the same headers that the "normal" `textpattern.js` and quickly cached. I have not noticed any persistent performance impact of the integrity checkup. If we change `textpattern.js` and its integrity hash, the browser will complain about any outdated cache.

Please test how [http://www.iut-fbleau.fr/projet/etc/textpattern/server.php?file=textpattern.js](http://www.iut-fbleau.fr/projet/etc/textpattern/server.php?file=textpattern.js) requests are cached by your OS/browser. If everything is ok, we can merge it into 4.7, with an extra security bonus. Thanks.

**Comment by phiw13** [comment]
I ran a quick test of your integrity branch with Safari 11, and everything seems to work fine. `server.php` comes down with a 304 status. So that seems Ok to me.

I’m not sure what the deal is with your test page. Here Safari always return a 200 status, no cache, whereas Firefox gives a 304.

**Comment by bloatware** [comment]
Thanks @phiw13, in my tests it behaves similarly: 304 or 200 on direct access, but cached when requested through `<script />`, so that's fine. I have to test it yet, but if the cache is updated on integrity violation, then victory.

**Comment by makss** [comment]
> 304 or 200

As far as I remember the webserver do it themselves, looked at the log on their test domain and chose the code `304`.
```
127.0.0.1 - - [28/Nov/2017:21:46:39 +0200] "GET /textpattern/admin-themes/hive/assets/css/textpattern.min.css HTTP/1.1" 304
127.0.0.1 - - [28/Nov/2017:21:46:39 +0200] "GET /textpattern/vendors/jquery/jquery/jquery.js HTTP/1.1" 304
127.0.0.1 - - [28/Nov/2017:21:46:39 +0200] "GET /textpattern/vendors/jquery/jquery-ui/jquery-ui.js HTTP/1.1" 304
127.0.0.1 - - [28/Nov/2017:21:46:39 +0200] "GET /textpattern/vendors/blueimp/md5/md5.min.js HTTP/1.1" 304
127.0.0.1 - - [28/Nov/2017:21:46:39 +0200] "GET /textpattern/vendors/blueimp/fileupload/jquery.fileupload.min.js HTTP/1.1" 304
127.0.0.1 - - [28/Nov/2017:21:46:39 +0200] "GET /textpattern/textpattern.js HTTP/1.1" 304
127.0.0.1 - - [28/Nov/2017:21:46:39 +0200] "GET /textpattern/admin-themes/hive/assets/css/textpattern.min.css HTTP/1.1" 304
127.0.0.1 - - [28/Nov/2017:21:46:39 +0200] "GET /textpattern/admin-themes/hive/assets/js/main.min.js HTTP/1.1" 304
127.0.0.1 - - [28/Nov/2017:21:46:39 +0200] "GET /textpattern/vendors/PrismJS/prism/prism.js HTTP/1.1" 304
127.0.0.1 - - [28/Nov/2017:21:46:39 +0200] "GET /textpattern/admin-themes/hive/assets/img/sprite.svg HTTP/1.1" 304
```

**Comment by makss** [comment]
I pressed the wrong button.  ;-)  Reopen

**Comment by bloatware** [comment]
You can easily simulate an update in [integrity](https://github.com/textpattern/textpattern/tree/integrity) branch now. Modify `textpattern.js` and see how your (modern!) browser reacts. If the file is cached, it should complain and download a fresh version on reload.

**Comment by Bloke** [comment]
Integrity is an ingenious idea. Love it, thanks Oleg.

Speaking of which, we need to add all the new classes to `checksums.txt` at some point... maybe when the `themes` branch lands.

**Comment by bloatware** [comment]
Should we replace md5 checksums with integrity hashes in `checksums.txt` and switch to JSON for easier hash extraction? Or introduce a new `integrity.json` file, or store hashes directly in `server.php`? Is there a way Github can fill hashes automatically?

**Comment by phiw13** [comment]
> You can easily simulate an update in integrity branch now. Modify textpattern.js and see how your (modern!) browser reacts. If the file is cached, it should complain and download a fresh version on reload.

Seems to work nicely with Safari 11, first check `server.php` returns a 304 (thing was cached), then after modifying `textpattern.js` I got a 200 and the browser fetched the whole file; a few clicks around the admin panels with a status 304. For kicks I modified the file again with an unclosed comment (!). everything went bonkers so surely it got a fresh file… That will good for smartphone users where a force reload is much more difficult.

**Comment by bloatware** [comment]
Ok, let us call it working overkill and postpone the decision to 4.8. A [version check](https://github.com/textpattern/textpattern/commit/8eecdbe881a57e3aef0036768f0c49ff4905054f) should suffice atm (a better wording is probably needed).

**Comment by Bloke** [comment]
@bloatware since the tests appeared to be positive, should we just check the integrity branch is up to date then merge it into dev?

**Comment by bloatware** [comment]
I've updated `integrity` branch, it seems to work. I'm not sure `css` files are served with integrity check too, will check.
